qrexec/tools/qrexec_policy_daemon.py

```diff
@@ -99,6 +106,114 @@ async def handle_client_connection(log, policy_cache,
+                      service_name, len(untrusted_data))
+            return
+        remote_domain = '@unknown'
+        # Qrexec guarantees that this space will be present
+        qrexec_command_with_arg, untrusted_data = untrusted_data.split(b' ', 1)
+        # Same with this NUL character
+        trusted_call_info, untrusted_data = untrusted_data.split(b'\0', 1)
+        # This space will only be present in dom0
+        remote_domain = trusted_call_info.split(b' ', 1)[0].decode('ascii', 'strict')
+        ### SANITIZE BEGIN
+        try:
+            invoked_service, service_queried = qrexec_command_with_arg.split(b'+', 1)
+        except ValueError:
+            log.warning('%s requires an argument (the service to query), but'
+                        '%s did not provide one', service_name, remote_domain)
+            return
+        if invoked_service != service_name:
+            # This is an error because qrexec should forbid this.
+            log.error('%s invoked with incorrect name %s by %s',
+                      service_name, invoked_service, remote_domain)
+            return
+
+        if not service_queried:
+            log.warning('%s: %s used the empty string as a service name',
+                        service_name, remote_domain)
+            return
+        if len(untrusted_data) > 63:
+            log.warning('%s: %s sent %d bytes, but the limit is 63',
+                        service_name, remote_domain, len(untrusted_data))
+            return
+        try:
+            untrusted_source, untrusted_target = untrusted_data.split(b'\0', 1)
+
+            # Check that qube name lengths are reasonable
+            # pylint: disable=superfluous-parens
+            if not (1 <= len(untrusted_target) <= 31):
+                log.warning('%s: %s sent a target domain name of length %d, but the limit is 31',
+                            service_name, remote_domain, len(untrusted_target))
+                return
+            # pylint: disable=superfluous-parens
+            if not (1 <= len(untrusted_source) <= 31):
+                log.warning('%s: %s sent a source domain name of length %d, but the limit is 31',
+                            service_name, remote_domain, len(untrusted_source))
+                return
+            untrusted_source = untrusted_source.decode('ascii', 'strict')
+            untrusted_target = untrusted_target.decode('ascii', 'strict')
+
+            # these throw exceptions if the domain name is not valid
+            sanitize_domain_name(untrusted_source, True)
+            sanitize_domain_name(untrusted_target, True)
+            ### SANITIZE END
+            source, intended_target = untrusted_source, untrusted_target
+        except (ValueError, UnicodeError) as e:
+            log.warning('%s: invalid data from qube %s', service_name, remote_domain)
+            return
+
+        if check_gui:
+            system_info = get_system_info()['domains']
```

Comment:
This makes the info loaded twice for a single request. I'm not sure what's the best solution, but some is to add `system_info` keyword argument (defaulting to `None`) to the `handle_request` function and make it load the info only if not already provided.